extras/rebalance/directory-rebalance.py

```diff
@@ -0,0 +1,203 @@
+        with open(self.index) as f:
+            sample_size = 1024*1024
+            sample_files = 100
+            i = 1
+            for line in f:
+                file_path, _, file_size = line.rpartition('-')
+                file_size = int(file_size)
+                self.migrate_with_stats(file_path, file_size)
+                if self.migrated_size > i*sample_size and self.migrated_files > i*sample_files:
+                    i += 1
+                    now = time.perf_counter()
+                    duration = now - self.rebalance_start
+                    if duration != 0.0:
+                        speed = self.migrated_size/duration
+                        migration_fraction = self.migrated_size/self.total_size
+                        eta = ((self.expected_total_size - self.total_size)/speed)*migration_fraction
+                        progress(self.total_size, self.expected_total_size, "ETA: {}".format(time_fmt(eta)))
+
+        self.exit()
+
+    #For each file in the directory recursively, writes <file-path>-<size> to index file
+    def generate_rebalance_file_index(self):
+        with open(self.index, 'w') as file_index:
+            total_size = 0
+            for d, _, files in os.walk(self.path):
+                for f in files:
+                    try:
+                        path = os.path.join(d, f)
+                        size = os.stat(path).st_size
+                        file_index.write(path+'-'+str(size)+'\n')
+                        self.expected_total_size += size
+                        self.expected_total_files += 1
+                        crawl_progress(self.expected_total_files, size_fmt(self.expected_total_size))
+                    except OSError as err:
+                        progress_done()
+                        print("OS error: {}".format(err))
+        progress_done()
+
+    #Stops the progress printing and prints stats collected so far
+    def exit(self):
+        progress_done()
+        rebalance_end = time.perf_counter()
+        self.duration = rebalance_end - self.rebalance_start
+
+        if self.total_files != 0:
+            print("Migrated {} / {} files [{:.2%}]".format(self.migrated_files, self.total_files, self.migrated_files/self.total_files))
+            if self.total_size != 0:
+                print("Migrated {} / {} data [{:.2%}]".format(size_fmt(self.migrated_size), size_fmt(self.total_size), self.migrated_size/self.total_size))
+        print("Run time: {}".format(time_fmt(self.duration)))
+        print("Time spent in migration: {} [{:.2%}]".format(time_fmt(self.migration_duration), self.migration_duration/self.duration))
+        print("Time spent in skipping: {} [{:.2%}]".format(time_fmt(self.skipped_migration_duration), self.skipped_migration_duration/self.duration))
+
+#/proc/mounts has active mount information. It checks that the given path is
+#mounted on glusterfs
+def check_glusterfs_supported_path(p):
+    abs_path = os.path.abspath(p)
```

Comment:
I think `realpath` is better than `abspath` in this case.

Reply:
Ah! didn't know about this subtle difference. Will change that.